File: sdks/swift/Sources/ExampleAppScannerSDK/ScannerClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Official Swift client for the example-app scanner API.
///
///   let client = ScannerClient(baseURL: URL(string: "https://app.example.com")!)
///   let pair = try await client.pair(.init(code: code, deviceLabel: "John iPhone"))
///   // Persist `pair.token` in Keychain.
///
///   let result = try await client.scan(.init(payload: scannedQr))
///   if result.wasAdmitted { admit(result.ticket) }
public actor ScannerClient {

    private let baseURL: URL
    private var token: String?
    private let session: URLSession
    private let timeout: TimeInterval
    private let throwOnDeny: Bool

    public init(
        baseURL: URL,
        token: String? = nil,
        session: URLSession = .shared,
        timeoutSeconds: TimeInterval = 10,
        throwOnDeny: Bool = true
    ) {
        self.baseURL = baseURL
        self.token = token
        self.session = session
        self.timeout = timeoutSeconds
        self.throwOnDeny = throwOnDeny
    }

    public var currentToken: String? { token }

    public func setToken(_ token: String?) { self.token = token }

    // MARK: - Endpoints

    public func pair(_ req: PairRequest) async throws -> PairResponse {
        let response: PairResponse = try await request("POST", "/api/v1/scanning/pair", body: req, requiresAuth: false)
        self.token = response.token
        return response
    }

    public func scan(_ req: ScanRequest) async throws -> ScanResult {
        do {
            let result: ScanResult = try await request("POST", "/api/v1/scanning/scan", body: req)
            try assertNotDenied(result)
            return result
        } catch let ScannerSDKError.apiError(status, _, raw) where status == 409 {
            // The deny path returns 409 with a ScanResult body. Unwrap.
            guard let raw = raw, let result = try? decoder.decode(ScanResult.self, from: raw) else {
                throw ScannerSDKError.apiError(status: status, code: "http_error", raw: raw)
            }
            try assertNotDenied(result)
            return result
        }
    }

    public func batchScan(_ scans: [ScanRequest]) async throws -> [ScanResult] {
        precondition(scans.count > 0, "Batch must be non-empty")
        precondition(scans.count <= 100, "Batch must be <= 100 scans")
        struct Body: Codable { let scans: [ScanRequest] }
        struct Wrap: Codable { let results: [ScanResult] }
        let w: Wrap = try await request("POST", "/api/v1/scanning/scan/batch", body: Body(scans: scans))
        return w.results
    }

    public func lookup(_ payload: String) async throws -> ScanResult {
        let path = "/api/v1/scanning/tickets/\(payload.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? payload)"
        return try await request("GET", path)
    }

    public func heartbeat(_ req: HeartbeatRequest = .init()) async throws -> HeartbeatResponse {
        try await request("POST", "/api/v1/scanning/heartbeat", body: req)
    }

    // MARK: - Internals

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    private func assertNotDenied(_ result: ScanResult) throws {
        if throwOnDeny && result.verdict == .deny {
            throw ScannerSDKError.scanDenied(reasonCode: result.reasonCode, flags: result.flags)
        }
    }

    private func request<R: Decodable>(
        _ method: String,
        _ path: String,
        body: Encodable? = nil,
        requiresAuth: Bool = true
    ) async throws -> R {
        guard !requiresAuth || token != nil else {
            throw ScannerSDKError.notPaired
        }

        var urlRequest = URLRequest(url: baseURL.appendingPathComponent(path))
        urlRequest.httpMethod = method
        urlRequest.timeoutInterval = timeout
        urlRequest.setValue("application/json", forHTTPHeaderField: "Accept")
        urlRequest.setValue("ExampleAppScannerSDK-Swift/1.0", forHTTPHeaderField: "User-Agent")
        if requiresAuth, let token { urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
            urlRequest.httpBody = try encoder.encode(AnyEncodable(body))
        }

        let (data, response) = try await session.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse else {
            throw ScannerSDKError.transport("Non-HTTP response")
        }

        guard (200..<300).contains(http.statusCode) else {
            let envelope = try? decoder.decode(ErrorEnvelope.self, from: data)
            throw ScannerSDKError.apiError(
                status: http.statusCode,
                code: envelope?.error ?? "http_error",
                raw: data
            )
        }

        return try decoder.decode(R.self, from: data)
    }
}

// MARK: - Error type

public enum ScannerSDKError: Error, Sendable {
    case notPaired
    case transport(String)
    case apiError(status: Int, code: String, raw: Data?)
    case scanDenied(reasonCode: String?, flags: [FraudFlag])
}

struct ErrorEnvelope: Decodable {
    let error: String?
    let message: String?
}

/// Type erasure so we can pass any `Encodable` as `body` and still
/// hand a concrete type to the encoder.
private struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void
    init(_ wrapped: Encodable) {
        self._encode = wrapped.encode
    }
    func encode(to encoder: Encoder) throws {
        try _encode(encoder)
    }
}

File: sdks/kotlin/src/main/kotlin/com/example_app/scanner/ScannerClient.kt
package com.example_app.scanner

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Official Android / Kotlin / Server-side JVM client for the
 * example-app scanner API.
 *
 *   val client = ScannerClient("https://app.example.com")
 *   val pair = client.pair(PairRequest(code = code, deviceLabel = "Operator phone"))
 *   client.token = pair.token  // persist via EncryptedSharedPreferences
 *
 *   val result = client.scan(ScanRequest(payload = scannedQr))
 *   if (result.wasAdmitted) admit(result.ticket)
 */
class ScannerClient @JvmOverloads constructor(
    private val baseUrl: String,
    var token: String? = null,
    private val httpClient: OkHttpClient = defaultClient(),
    private val throwOnDeny: Boolean = true,
) {

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
    }
    private val jsonMedia = "application/json; charset=utf-8".toMediaType()

    suspend fun pair(req: PairRequest): PairResponse {
        val response: PairResponse = request("POST", "/api/v1/scanning/pair", req, requiresAuth = false)
        this.token = response.token
        return response
    }

    suspend fun scan(req: ScanRequest): ScanResult {
        val result = runCatching { request<ScanResult>("POST", "/api/v1/scanning/scan", req) }
            .recoverCatching { e ->
                // 409 on deny carries a ScanResult body. Re-decode it.
                if (e is ScannerApiException && e.status == 409 && e.rawBody != null) {
                    json.decodeFromString<ScanResult>(e.rawBody)
                } else throw e
            }
            .getOrThrow()

        if (throwOnDeny && result.verdict == ScanVerdict.DENY) {
            throw ScanDeniedException(result.reasonCode, result.flags)
        }
        return result
    }

    suspend fun batchScan(scans: List<ScanRequest>): List<ScanResult> {
        require(scans.isNotEmpty()) { "Batch must be non-empty" }
        require(scans.size <= 100) { "Batch must be <= 100 scans" }
        @kotlinx.serialization.Serializable
        data class Body(val scans: List<ScanRequest>)

        @kotlinx.serialization.Serializable
        data class Wrap(val results: List<ScanResult>)

        return request<Wrap>("POST", "/api/v1/scanning/scan/batch", Body(scans)).results
    }

    suspend fun lookup(payload: String): ScanResult {
        val encoded = URLEncoder.encode(payload, "UTF-8")
        return request("GET", "/api/v1/scanning/tickets/$encoded")
    }

    suspend fun heartbeat(req: HeartbeatRequest = HeartbeatRequest()): HeartbeatResponse =
        request("POST", "/api/v1/scanning/heartbeat", req)

    private suspend inline fun <reified R> request(
        method: String,
        path: String,
        body: Any? = null,
        requiresAuth: Boolean = true,
    ): R = withContext(Dispatchers.IO) {
        if (requiresAuth && token == null) throw NotPairedException()

        val builder = Request.Builder()
            .url(baseUrl.trimEnd('/') + path)
            .header("Accept", "application/json")
            .header("User-Agent", "ExampleAppScannerSDK-Kotlin/1.0")

        if (requiresAuth) token?.let { builder.header("Authorization", "Bearer $it") }

        when {
            body == null && method == "GET" -> builder.get()
            body == null -> builder.method(method, ByteArray(0).toRequestBody(jsonMedia))
            else -> {
                val payload = json.encodeToString(serializerOf(body), body)
                builder.method(method, payload.toRequestBody(jsonMedia))
            }
        }

        val raw = httpClient.newCall(builder.build()).execute().use { resp ->
            val bodyStr = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) {
                val code = runCatching { json.decodeFromString<ErrorEnvelope>(bodyStr).error }
                    .getOrNull() ?: "http_error"
                throw ScannerApiException(resp.code, code, rawBody = bodyStr)
            }
            bodyStr
        }
        json.decodeFromString<R>(raw)
    }

    @Suppress("UNCHECKED_CAST")
    private fun serializerOf(body: Any) = json.serializersModule.serializer(body::class.java)
        as kotlinx.serialization.KSerializer<Any>

    companion object {
        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build()
    }
}

@kotlinx.serialization.Serializable
private data class ErrorEnvelope(val error: String? = null, val message: String? = null)

class NotPairedException : RuntimeException("Not paired — call pair() first.")
class ScannerApiException(val status: Int, val code: String, val rawBody: String? = null) :
    RuntimeException("HTTP $status — $code")
class ScanDeniedException(val reasonCode: String?, val flags: List<FraudFlag>) :
    RuntimeException("Scan denied: ${reasonCode ?: "unknown"}")
